proposals/ref-readonly-parameters.md

```diff
@@ -0,0 +1,112 @@
+| `ref`                | Allowed         | **Allowed**              | **Warning**    | Error           |
+| `in`                 | Error           | **Allowed**              | Allowed        | Error           |
+| `out`                | Error           | **Error**                | Error          | Allowed         |
+| No annotation        | Error           | **Warning**              | Allowed        | Error           |
+
+Note that there is one change to the existing rules: `in` parameter with `ref` callsite annotation produces a warning instead of an error.
+
+| Value kind | `ref` parameter | `ref readonly` parameter | `in` parameter | `out` parameter |
+|------------|-----------------|--------------------------|----------------|-----------------|
+| rvalue     | Error           | **Warning**              | Allowed        | Error           |
+| lvalue     | Allowed         | **Allowed**              | Allowed        | Allowed         |
+
+## Motivation
+[motivation]: #motivation
+
+C# 7.2 [introduced `in` parameters](https://github.com/dotnet/csharplang/blob/c8c1615fcad4ca016a97b7260ad497aad53ebc78/proposals/csharp-7.2/readonly-ref.md#solution-in-parameters) as a way to pass readonly references.
+`in` parameters allow both lvalues and rvalues and can be used without any annotation at the callsite.
+However, APIs which capture or return references from their parameters would like to disallow rvalues and also enforce some indication at the callsite that a reference is being captured.
+`ref readonly` parameters are ideal in such cases as they warn if used with rvalues or without any annotation at the callsite.
+
+Furthermore, there are APIs that need only read-only references but use
+- `ref` parameters since they were introduced before `in` became available and changing to `in` would be a source and binary breaking change, e.g., `QueryInterface`, or
+- `in` parameters to accept readonly references even though passing rvalues to them doesn't really make sense, e.g., `ReadOnlySpan<T>..ctor(in T value)`, or
+- `ref` parameters to disallow rvalues even though they don't mutate the passed reference, e.g., `Unsafe.IsNullRef`.
+
+These APIs could migrate to `ref readonly` parameters without breaking users.
+For details on binary compatibility, see the proposed [metadata encoding](#metadata).
+Specifically, changing
+- `ref` → `ref readonly` would only be a binary breaking change for virtual methods,
+- `ref` → `in` would also be a binary breaking change for virtual methods, but not a source breaking change (because the rules change to only warn for `ref` arguments passed to `in` parameters),
+- `in` → `ref readonly` would not be a breaking change (but no callsite annotation or rvalue would result in a warning).
+
+In the opposite direction, changing
+- `ref readonly` → `ref` would be potentially a source breaking change (unless only `ref` callsite annotation was used), and a binary breaking change for virtual methods,
```

Comment:
It would be good to mention here that it would be a source breaking change also since it would now disallow readonly refs to be passed to the method.